Remove a layer of caching for series

cevian opened the pull request:
Previously, series were stored in two separate caches:
1) the label interner
2) the handlers series->series id cache

This commit unifies that into 1 cache (the interner) which now
stores both the series and it's id.

The listener for epoch changes has also now moved from a per-handler
thing to a more singleton ingestor proc. There is no reason to
have many epoch change listeners since there is only one epoch.

blagojts reviewed (approved):
Looks a lot simpler now

Harkishen-Singh commented on pkg/pgmodel/model/labels.go:
```diff
@@ -9,23 +9,50 @@ import (
+// to the PostgreSQL type in the series table (currently BIGINT).
+type SeriesID int64
+
+const unsetSeriesID = -1
```
can we have this as `defaultSeriesID`? unsetSeriesID makes me sound like a function that unsets the seriesID, which is unrelated.

blagojts replied:
It's an invalid seriesID, shouldn't be called default

Harkishen-Singh replied:
I agree, but something like initSeriesID then?

cevian replied:
renaming to invalidSeriesID

Harkishen-Singh commented on pkg/pgmodel/model/labels.go:
```diff
@@ -9,23 +9,50 @@ import (
+	return strconv.FormatInt(int64(s), 10)
+}
+
+//Epoch represents the series epoch
+type SeriesEpoch int64
```
Sorry, but whenever I read the epoch thing, I still get confused even today. Can we please add a good explanation somewhere here, on why is epoch needed and what's it all about? The importance of this would increase more as we have more contributors.

blagojts replied:
There is a readme https://github.com/timescale/promscale/blob/master/pkg/pgmodel/epochs.readme.md but maybe we need to add more details to it

cevian replied:
^ what blagoj said

Harkishen-Singh commented on pkg/pgmodel/model/labels.go:
```diff
@@ -9,23 +9,50 @@ import (
+//Epoch represents the series epoch
+type SeriesEpoch int64
+
+const UnsetSeriesEpoch = -1
```
`defaultSeriesEpoch`?

Harkishen-Singh commented on pkg/pgmodel/model/labels.go:
```diff
@@ -130,10 +157,13 @@ func LabelProtosToLabels(labelPairs []prompb.Label) (*Labels, string, error) {
+			Names:    make([]string, len(labelPairs)),
+			Values:   make([]string, len(labelPairs)),
+			str:      str,
+			seriesID: unsetSeriesID,
```
A thing for discussion. We create a new Labels object and then store in the cache with the seriesID being `-1`. Then, every time before insertion, we find out that seriesID (from table if its new) and replace -1 with the ID it deserves. Do we make sure that the cache is updated with that finding? If not, we are wasting a lot a CPU and also trips to DB for no reason.

Harkishen-Singh replied:
If we do not update the cache with new series ID, then can we do the findings of a seriesID here and insert directly in the cache with a non-default (non `-1`) seriesID, and it again saves a complete loop from the cache which we would do later on to fill the seriesID to this `-1`.

blagojts replied:
1. Yes cache is updated, we most likely have an integration test for that
2. If you set the series id here as reported by the db, you waste the batching done currently where all the missing series id being retrieved in one query

cevian replied:
Yes this way is done for batching. It is updated when set because this is the same object in the cache

Harkishen-Singh commented on pkg/pgmodel/ingestor/ingestor.go:
```diff
@@ -92,9 +93,8 @@ func (i *DBIngestor) parseData(tts []prompb.TimeSeries, req *prompb.WriteRequest
-			SeriesID: -1, // sentinel marking the seriesId as unset
-			Samples:  t.Samples,
+			Labels:  seriesLabels,
+			Samples: t.Samples,
```
(not in this PR) Instead of making a new local allocation, should SamplesInfo struct just extend the `promb.Timeseries` with just our `Labels`? This will save from local allocation, especially in high ingestion with more samples count.

blagojts replied:
I think extensions are just syntactic sugar for writing 
```
type myType{
  typeImExtending
  somethingElse OfType
}
```
The alloc will happen

Harkishen-Singh replied:
I mean, since the incoming ts are already allocated, so something like

```
type PSSeries struct {
  *promb.Samples
  labels *Labels
}
```

then using it as

```
&PSSeries{&t.Samples, seriesLabels}
```

will this be any good?

Harkishen-Singh replied:
I wrote this assuming it was all pointers coming, but now I see them as not. So, just in case.

cevian replied:
Also t.Samples is a slice so what we are doing is just copyng a pointer

Harkishen-Singh commented on pkg/pgmodel/model/labels.go:
```diff
@@ -164,18 +194,24 @@ func (l *Labels) Equal(b *Labels) bool {
+	switch l.seriesID {
+	case unsetSeriesID:
+		return 0, 0, fmt.Errorf("Series id not set")
+	case 0:
```
Is this possible? (for sake of knowing more)

cevian replied:
Not unless it's unset(programming error)

Harkishen-Singh replied:
If that has to be programming error, then we should have a explicit check for this in tests, so that it can be caught before merging.

cevian commented:
Changes up to and including "Optimize memory: unset series labels and names when setting id" save memory by 33%:
Representative heap profile before:
```
(pprof) top
Showing nodes accounting for 422.28MB, 82.02% of 514.87MB total
Dropped 56 nodes (cum <= 2.57MB)
Showing top 10 nodes out of 89
      flat  flat%   sum%        cum   cum%
   89.01MB 17.29% 17.29%   214.66MB 41.69%  github.com/timescale/promscale/pkg/pgmodel/model.LabelProtosToLabels
   82.53MB 16.03% 33.32%    82.53MB 16.03%  strings.(*Builder).grow (inline)
      66MB 12.82% 46.14%       66MB 12.82%  github.com/timescale/promscale/pkg/prompb.(*Label).Unmarshal
   52.52MB 10.20% 56.34%   118.52MB 23.02%  github.com/timescale/promscale/pkg/prompb.(*TimeSeries).Unmarshal
   46.09MB  8.95% 65.29%    46.09MB  8.95%  github.com/timescale/promscale/pkg/pgmodel/ingestor.(*pendingBuffer).addReq
   24.28MB  4.72% 70.00%    24.28MB  4.72%  github.com/timescale/promscale/pkg/pgmodel/ingestor.(*pgxInserter).getMetricInserter
   19.81MB  3.85% 73.85%    36.63MB  7.11%  sync.(*Map).LoadOrStore
   15.13MB  2.94% 76.79%    26.24MB  5.10%  github.com/timescale/promscale/pkg/pgmodel/ingestor.(*insertHandler).setSeriesIds
   13.81MB  2.68% 79.47%    13.81MB  2.68%  sync.(*Map).dirtyLocked
   13.10MB  2.54% 82.02%    13.10MB  2.54%  github.com/jackc/chunkreader/v2.(*ChunkReader).newBuf (inline)
```
and after: 
```
Showing nodes accounting for 294.52MB, 86.59% of 340.13MB total
Dropped 57 nodes (cum <= 1.70MB)
Showing top 10 nodes out of 80
      flat  flat%   sum%        cum   cum%
   66.52MB 19.56% 19.56%    66.52MB 19.56%  strings.(*Builder).grow
   58.15MB 17.10% 36.65%    58.15MB 17.10%  github.com/timescale/promscale/pkg/clockcache.(*Cache).Reset
   45.39MB 13.34% 50.00%    45.39MB 13.34%  github.com/timescale/promscale/pkg/pgmodel/ingestor.(*pendingBuffer).addReq
   44.51MB 13.08% 63.08%    44.51MB 13.08%  github.com/timescale/promscale/pkg/pgmodel/model.NewSeries
   26.39MB  7.76% 70.84%    26.39MB  7.76%  github.com/timescale/promscale/pkg/pgmodel/ingestor.(*pgxInserter).getMetricInserter
   17.01MB  5.00% 75.84%    32.51MB  9.56%  github.com/timescale/promscale/pkg/prompb.(*TimeSeries).Unmarshal
   15.50MB  4.56% 80.40%    15.50MB  4.56%  github.com/timescale/promscale/pkg/prompb.(*Label).Unmarshal
       9MB  2.65% 83.04%        9MB  2.65%  fmt.Sprintf
    6.55MB  1.93% 84.97%     6.55MB  1.93%  github.com/jackc/chunkreader/v2.(*ChunkReader).newBuf (inline)
    5.51MB  1.62% 86.59%    31.88MB  9.37%  github.com/timescale/promscale/pkg/pgmodel/ingestor.doInsert
```
```

Harkishen-Singh commented on pkg/pgmodel/cache/series_cache.go:
```diff
@@ -2,70 +2,81 @@
-	str        string
+const DefaultSeriesCacheSize = 500000
+
+type SeriesCache interface {
+	Reset()
+	GetSeriesFromProtos(labelPairs []prompb.Label) (series *model.Series, metricName string, err error)
+	Len() int
+	Cap() int
+}
```
This interface is used only once (which does not justify its need).
Can we not have a generic interface, something like 

```
type Cache interface {
  Len() int
  Cap() int
  Reset()
}
```
And this should be inside `pkg/clockcache` and then imported from there to every other place. This would be awesome for the cache implementation. Else, I see 3 different kinds of interfaces - `SeriesCache`, `MetricCache`, `LabelsCache` and all have a singlar usage (except MetricCache, which can be inside the Cache interface as I said), which is not great.

cevian replied:
We have already seperate interfaces for MetricCache and and LabelCache, so I'm just following an established convention. If we want to change, let's discuss but not for this PR

Harkishen-Singh replied:
We should change that, sure for the other PR, but then we should add a note in the project board.

blagojts replied:
The MetricCache interface has methods that simplify calling the underlying clock cache,
so we don't have to convert string to []interface{}
```
type MetricCache interface {
	Get(metric string) (string, error)
	Set(metric string, tableName string) error
	Len() int
	Cap() int
}

type LabelsCache interface {
	GetValues(keys []interface{}, valuesOut []interface{}) (numFound int)
	InsertBatch(keys []interface{}, values []interface{}) (numInserted int)
	Len() int
	Cap() int
}
```
And the SeriesCache will have a `GetSeriesFromProtos(labelPairs []prompb.Label) (series *model.Series, metricName string, err error)` that does not apply to the other caches, we can put the Len, Cap and Reset methods to an interface in the clockcache package, but we either keep these 3 (while extending the clockcache.Cache interface) , or define a new SuperCacheInterface that has methods for all of them, or write new wrappers around the single interface that will hold the logic for the now distinct methods

blagojts replied:
We can end up with 
```go
type SeriesCache interface {
	clockcache.CacheInterface
	GetSeriesFromProtos(labelPairs []prompb.Label) (series *model.Series, metricName string, err error)
}

type seriesCacheImpl struct {
	*clockcache.CacheStruct
}

func GetSeriesFromProtos(labelPairs []prompb.Label) (series *model.Series, metricName string, err error){
  panic("implement me")
}
```

Harkishen-Singh commented on pkg/pgmodel/ingestor/inserter.go:
```diff
@@ -125,9 +166,8 @@ func (p *pgxInserter) InsertData(rows map[string][]model.SamplesInfo) (uint64, e
-		// just a channel insert
-		p.insertMetricData(metricName, data, workFinished, errChan)
+		// the following is usually non-blocking, just a channel insert
+		p.getMetricInserter(metricName) <- insertDataRequest{metric: metricName, data: data, finished: workFinished, errChan: errChan}
```
Should we be inserting a pointer? given that channels have limitations in speed for large data, which is why we pass address in prom-migrator as well.

cevian replied:
that's changed in PR 502

Harkishen-Singh commented on pkg/pgmodel/ingestor/inserter.go:
```diff
@@ -107,11 +108,11 @@ func (p *pgxInserter) refreshSeriesEpoch(existingEpoch model.SeriesEpoch) model.
 		// and continue execution
 		log.Error("msg", "error refreshing the series cache", "err", err)
```
(optional) I think we should always do bare logging from the parent, and from a child. So, this should return an error and `runSeriesEpochSync` would be responsible for logging it out.

Harkishen-Singh commented on pkg/pgmodel/model/series.go:
```diff
@@ -89,6 +115,8 @@ func (l *Series) GetSeriesID() (SeriesID, SeriesEpoch, error) {
 
 //note this has to be idempotent
 func (l *Series) SetSeriesID(sid SeriesID, eid SeriesEpoch) {
+	l.lock.Lock()
```
This is good.

Harkishen-Singh commented on pkg/pgmodel/cache/series_cache.go:
```diff
@@ -2,70 +2,81 @@
-	Values     []string
-	MetricName string
-	str        string
+const DefaultSeriesCacheSize = 500000
```
Can we please have a comment as to how setting this high helps us, based on the recent benchmarks?

Harkishen-Singh reviewed (changes requested):
Few comments, rest all looks good.

cevian commented on pkg/pgmodel/ingestor/handler.go:
```diff
@@ -147,118 +104,85 @@ func (h *insertHandler) setSeriesIds(sampleInfos []model.SamplesInfo) (string, E
 	}
 
 	if numSQLFunctionCalls != len(batchSeries) {
-		return "", epoch, fmt.Errorf("unexpected difference in numQueries and batchSeries")
+		return fmt.Errorf("unexpected difference in numQueries and batchSeries")
 	}
```
I don't think so. batchSeries is only appended to in line 127 so the continue in those two blocks should not effect the length of batchSeries.

cevian commented on pkg/pgmodel/model/series.go:
```diff
@@ -0,0 +1,122 @@
+	return l.str == b.str
+}
+
+func (l *Series) isSeriesIDSetNoLock() bool {
+	return l.seriesID != invalidSeriesID
+}
```
we need this for the NameValues function which already takes it's own lock. Note this is a private function

cevian commented on pkg/pgmodel/ingestor/handler.go:
```diff
@@ -120,21 +80,18 @@ func (h *insertHandler) flushPending() {
-	}
-
-	seriesToInsert := make([]*model.SamplesInfo, 0, numMissingSeries)
+func (h *insertHandler) setSeriesIds(sampleInfos []model.SamplesInfo) error {
```
done. (in a more user-friendly error)

Harkishen-Singh reviewed (approved):
Looks good, but I am not sure if we should go with the merge, unless results from prom-migrator is clear.

Harkishen-Singh commented:
After careful evaluation of the changes in relation with the perf testing, I don't see any changes over here would slow down in the perf tests. Rather, things like getStr()/parseData() have actually speed up, which is obvious due to change in GetProtoToPRotoLabels.

However, for some reason, in each prom-migrator test, I find this branch lagging behind by at least 1 second. I don't have any proof from profiling, so for now, I think will have to take it as noise and move ahead.

Thank you.

blagojts commented on pkg/pgmodel/ingestor/inserter.go:
```diff
@@ -86,6 +94,42 @@ func (p *pgxInserter) runCompleteMetricCreationWorker() {
+	dbEpoch, err := p.getServerEpoch()
+	if err != nil {
+		// Trash the cache just in case an epoch change occurred, seems safer
+		p.scache.Reset()
```
Is it a good idea to store the epoch with each series so we can purge only data from an earlier epoch?

cevian replied:
Good idea but I think separate PR

blagojts reviewed (approved):
Looks great. Left one question re: cache resets, which if it makes sense might go in a separate PR
